examples/blame_example.ipynb

```diff
@@ -1,310 +1,240 @@
-    "scene_plotter.animate_scene(output_name=gif_name, ax=ax, \n",
-    "                            start_idx=max(0, collision[1][0] - 50), end_idx=collision[1][1],\n",
-    "                            numbers=list(collision[0]), direction_vec=False, velocity_vec=False,\n",
-    "                            plot_frame_number=True)"
-   ]
-  },
-  {
-   "cell_type": "code",
-   "execution_count": 12,
-   "metadata": {},
-   "outputs": [
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "L0ufxMO8NdvX"
+      },
+      "outputs": [],
+      "source": [
+        "collisions = []\n",
+        "collision_index = None\n",
+        "is_found_collision = False\n",
+        "\n",
+        "# Simulate with `initialize`, `drive` and `light` until there are collisions.\n",
+        "for _ in range(20): #Attempt 20 simulations looking for a collision\n",
+        "    light_response = iai.light(location=location)\n",
+        "\n",
+        "    response = iai.initialize(\n",
+        "        location=location,\n",
+        "        agent_count=15,\n",
+        "        get_birdview=True,\n",
+        "        traffic_light_state_history=[light_response.traffic_lights_states]\n",
+        "    )\n",
+        "    agent_attributes = response.agent_attributes\n",
+        "    scene_plotter.initialize_recording(response.agent_states,\n",
+        "                                       agent_attributes=agent_attributes,\n",
+        "                                       traffic_light_states=light_response.traffic_lights_states)\n",
+        "\n",
+        "    agent_state_history = []\n",
+        "    traffic_light_state_history = []\n",
+        "\n",
+        "    # 20-second scene\n",
+        "    for t in range(100):\n",
+        "        light_response = iai.light(location=location, recurrent_states=light_response.recurrent_states)\n",
+        "        response = iai.drive(\n",
+        "            location=location,\n",
+        "            agent_attributes=agent_attributes,\n",
+        "            agent_states=response.agent_states,\n",
+        "            recurrent_states=response.recurrent_states,\n",
+        "            get_birdview=False,\n",
+        "            traffic_lights_states=light_response.traffic_lights_states,\n",
+        "            get_infractions=True,\n",
+        "            random_seed=1\n",
+        "        )\n",
+        "        scene_plotter.record_step(response.agent_states, traffic_light_states=light_response.traffic_lights_states)\n",
+        "        agent_state_history.append(response.agent_states)\n",
+        "        traffic_light_state_history.append(light_response.traffic_lights_states)\n",
+        "\n",
+        "        for agent, infractions in enumerate(response.infractions):\n",
+        "            if infractions.collisions: collisions.append(agent)\n",
+        "            if len(collisions) == 2 and not is_found_collision:\n",
```

Comment:
I don't think the code logic here matches the colliding agent pairs correctly, especially when there are more than one collisions in the same frame. 

You can copy the [function](https://github.com/inverted-ai/driving-models/blob/5559f4c53580a27692bdbec7c2cd2769b63eeeca/api_server/app/utils.py#L448) in the server side and change accordingly based on your needs. The idea is to match the colliding agent pair based on the overlap.

Reply:
That can be done, this will be temporary until we create a collision detection feature within the invertedai library and should be replaced then.

Reply:
Now that I've looked into this a bit more, I am realizing the complications. This repo is public facing but the function you linked is from a private repo thus cannot be included. We do not have a public facing collision detection feature except the one in drive that is limited. I think the solution is that there are a selection of logs that can be played instead of generating a collision scenario live. What are your thoughts on this @zhangk1551 since it will change this example quite a lot?

Reply:
For this particular [function](https://github.com/inverted-ai/driving-models/blob/5559f4c53580a27692bdbec7c2cd2769b63eeeca/api_server/app/utils.py#L448), it's standalone and doesn't rely on other functions in the private repo. We can move it to the public client repo directly.

Reply:
A collision detection function has been added to the demo notebook.